zzre/game/systems/dialog/DialogChestPuzzle.cs

```diff
@@ -0,0 +1,209 @@
+            entity.Dispose();
+    }
+
+    protected override void HandleOpen(in messages.DialogChestPuzzle message)
+    {
+        message.DialogEntity.Set(components.DialogState.ChestPuzzle);
+
+        World.Publish(new messages.DialogResetUI(message.DialogEntity));
+        var uiEntity = World.CreateEntity();
+        uiEntity.Set(new components.Parent(message.DialogEntity));
+
+        preload.CreateDialogBackground(uiEntity, animateOverlay: true, out var bgRect, opacity: 1f);
+
+        uiEntity.Set(new components.DialogChestPuzzle{
+            DialogEntity = message.DialogEntity,
+            Size = message.Size,
+            LabelExit = message.LabelExit,
+            NumAttempts = 0,
+            BoardState = InitBoardState(message.Size),
+            BgRect = bgRect
+        });
+        ref var puzzle = ref uiEntity.Get<components.DialogChestPuzzle>();
+
+        preload.CreateLabel(uiEntity)
+            .With(puzzle.BgRect.Min + new Vector2(20, 20))
+            .With(preload.Fnt001)
+            .WithText(db.GetText(UIDBoxOfTricks).Text)
+            .Build();
+
+        puzzle.Board = CreateBoard(uiEntity, ref puzzle);
+        puzzle.Attempts = CreateAttempts(uiEntity, ref puzzle);
+        puzzle.Action = preload.CreateSingleDialogButton(uiEntity, UIDCancel, IDCancel, puzzle.BgRect, buttonOffsetY: -45f);
+    }
+
+    private static bool[] InitBoardState(int size)
+    {
+        var board = new bool[size*size];
+        for (int i = 0; i < size*size; i++)
+            board[i] = Convert.ToBoolean(i % 2);
+        return board;
+    }
+
+    private DefaultEcs.Entity CreateAttempts(DefaultEcs.Entity parent, ref components.DialogChestPuzzle puzzle)
+    {
+        var entity = World.CreateEntity();
+        entity.Set(new components.Parent(parent));
+
+        preload.CreateLabel(entity)
+            .With(puzzle.BgRect.Min + new Vector2(25, 120))
+            .With(preload.Fnt000)
+            .WithText($"{db.GetText(UIDAttempts).Text}: {puzzle.NumAttempts}\n{db.GetText(UIDMinTries).Text}: {savegame.switchGameMinMoves}")
+            .WithLineHeight(15)
+            .Build();
+
+        return entity;
+    }
+
+    private DefaultEcs.Entity CreateBoard(DefaultEcs.Entity parent, ref components.DialogChestPuzzle puzzle)
+    {
+        var entity = World.CreateEntity();
+        entity.Set(new components.Parent(parent));
+
+        var offset = new Vector2(-1, -1) + new Vector2(-23, -23) * puzzle.Size;
+
+        for (int row = 0; row < puzzle.Size; row++)
+        {
+            for (int col = 0; col < puzzle.Size; col++)
+            {
+                var cell = row * puzzle.Size + col;
+                var IDCell = new components.ui.ElementId(cell);
+                var button = preload.CreateButton(entity)
+                    .With(IDCell)
+                    .With(offset + new Vector2(46 * col, 46 * row))
+                    .With(new components.ui.ButtonTiles(puzzle.BoardState[cell] ? 1 : 2))
```

Comment:
Similar to the attempts label you can update the board entities instead of recreating them. There is a rare feature of `ButtonTiles` where you can set/remove the `components.ui.Active` component to switch the shown tile IDs